sqlglot/dialects/oracle.py

```diff
@@ -282,3 +282,180 @@ def add_column_sql(self, expression: exp.AlterTable) -> str:
             if len(expression.args.get("actions", [])) > 1:
                 return f"ADD ({actions})"
             return f"ADD {actions}"
+
+
+def eliminate_join_marks(ast: exp.Expression) -> exp.Expression:
+    from sqlglot.optimizer.scope import build_scope
+
+    """Remove join marks from an expression
+    
+    SELECT * FROM a, b WHERE a.id = b.id(+)   
+    becomes:
+    SELECT * FROM a LEFT JOIN b ON a.id = b.id
+
+    - for each scope
+        - for each column with a join mark
+            - find the predicate it belongs to
+            - remove the predicate from the where clause
+            - convert the predicate to a join with the (+) side as the left join table
+            - replace the existing join with the new join
+
+    Args:
+        ast: The AST to remove join marks from
+
+    Returns:
+       The AST with join marks removed"""
+    scopes = build_scope(ast)
+    if not scopes:
+        return ast
+    for scope in scopes.traverse():
+        _eliminate_join_marks_from_scope(scope)
+    return ast
+
+
+def _update_from(
+    select: exp.Select,
+    new_join_dict: t.Dict[str, exp.Join],
+    old_join_dict: t.Dict[str, exp.Join],
+) -> None:
+    """If the from clause needs to become a new join, find an appropriate table to use as the new from.
+    updates select in place
```

Comment:
Can you explain the logic here a bit? Why would the `FROM` need to become a new join? Is this because you can have:

```sql
SELECT * FROM a, b WHERE a.x (+) = b.x
```

where `b` becomes the left relation in the `LEFT JOIN`?

Reply:
presently, the transformation is applied after the oracle parser. So
SELECT * FROM a,b WHERE a.x (+) = b.x 
becomes
SELECT * FROM a JOIN b WHERE a.x (x) = b.x
But the query should become
SELECT * FROM b LEFT JOIN a ...